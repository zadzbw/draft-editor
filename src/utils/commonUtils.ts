const quoteReg = /<blockquote>([\w\W]*?)<\/blockquote>/g;

/**
 * 将blockquote内的多个p标签，同blockquote一起，转换成多个blockquote标签
 * @param html
 * @return {string}
 */
export const transformBlockQuote = (html: string): string => {
  return html.replace(quoteReg, (match, p) => {
    return p.replace(/<p>/g, '<blockquote>').replace(/<\/p>/g, '</blockquote>');
  });
};

/**
 * 判断是否为mac平台
 * @return {boolean}
 */
export const isMac = (): boolean => {
  return window.navigator && /Mac/i.test(navigator.userAgent);
};

export const getShortcut = (shortcut: string): string => {
  if (isMac()) {
    return shortcut.replace(/\+/g, '');
  }
  return shortcut
    .replace('⌘', 'Ctrl')
    .replace('⇧', 'Shift')
    .replace('⌥', 'Alt');
};
