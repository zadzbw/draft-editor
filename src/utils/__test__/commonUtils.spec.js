import { transformBlockQuote, getShortcut } from '../commonUtils';

const ua = navigator.userAgent;

describe('commonUtils', () => {
  describe('transformBlockQuote', () => {
    it('should do nothing when has no blockquote ', () => {
      const html = '<p>test</p>';
      const result = transformBlockQuote(html);
      expect(result).toBe(html);
      expect(result).toMatchSnapshot();
    });

    it('should transform single blockquote correctly', () => {
      const html = '<blockquote><p>111</p><p>222<br>333</p></blockquote>';
      const result = transformBlockQuote(html);
      expect(result).toMatchSnapshot();
    });

    it('should transform multiple blockquote correctly', () => {
      const html = '<blockquote><p>111</p><p>222<br>333</p></blockquote><p>test</p><blockquote><p>444</p><p>555<br>666</p></blockquote>';
      const result = transformBlockQuote(html);
      expect(result).toMatchSnapshot();
    });

    it('should transform adjacent blockquote correctly', () => {
      const html = '<blockquote><p>111</p><p>222<br>333</p></blockquote><blockquote><p>444</p><p>555<br>666</p></blockquote>';
      const result = transformBlockQuote(html);
      expect(result).toMatchSnapshot();
    });
  });

  // 每个测试完成后重置ua
  // eslint-disable-next-line no-undef
  afterEach(() => {
    Object.defineProperty(navigator, 'userAgent', {
      writable: true,
      value: ua,
    });
  });

  describe('getShortcut', () => {
    it('should get shortcut in mac', () => {
      Object.defineProperty(navigator, 'userAgent', {
        writable: true,
        value:
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36',
      });
      const shortcut = '⌘+⌥+⇧+B';
      const result = getShortcut(shortcut);
      expect(result).toBe('⌘⌥⇧B');
      expect(result).toMatchSnapshot();
    });

    it('should get shortcut in windows', () => {
      const shortcut = '⌘+⌥+⇧+B';
      const result = getShortcut(shortcut);
      expect(result).toBe('Ctrl+Alt+Shift+B');
      expect(result).toMatchSnapshot();
    });
  });
});
