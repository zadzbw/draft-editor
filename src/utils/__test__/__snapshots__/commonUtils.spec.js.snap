// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commonUtils getShortcut should get shortcut in mac 1`] = `"⌘⌥⇧B"`;

exports[`commonUtils getShortcut should get shortcut in windows 1`] = `"Ctrl+Alt+Shift+B"`;

exports[`commonUtils transformBlockQuote should do nothing when has no blockquote  1`] = `"<p>test</p>"`;

exports[`commonUtils transformBlockQuote should transform adjacent blockquote correctly 1`] = `"<blockquote>111</blockquote><blockquote>222<br>333</blockquote><blockquote>444</blockquote><blockquote>555<br>666</blockquote>"`;

exports[`commonUtils transformBlockQuote should transform multiple blockquote correctly 1`] = `"<blockquote>111</blockquote><blockquote>222<br>333</blockquote><p>test</p><blockquote>444</blockquote><blockquote>555<br>666</blockquote>"`;

exports[`commonUtils transformBlockQuote should transform single blockquote correctly 1`] = `"<blockquote>111</blockquote><blockquote>222<br>333</blockquote>"`;
